File: backend/src/profile/profileService.ts
import { GithubRepo } from "@/common/interface";
import axios from "axios";
import { cache } from "@/common/cache";
import { ProfileData } from "@/common/interface";
import fs from "fs";
import path from "path";

const profileData: ProfileData = JSON.parse(fs.readFileSync(path.join(__dirname, "../../conf/portfolio.json"), "utf-8"))
const icons = fs.readdirSync(path.join(__dirname, "../../public/icons"));
const techStackList = icons.map(file => `/static/icons/${file}`);

export class ProfileService {
    private static readonly BASE_URL = "https://api.github.com";
    private static readonly USERNAME = "jung18";
    private static readonly GITHUB_TOKEN = process.env.GITHUB_TOKEN;
    private static readonly CACHE_KEY = "github_repos";

    public static getProfile() {
        return profileData;
    }

    public static getTechStackList() {
        return techStackList;
    }

    public static async getRepos() {
        let repos = this.getCache();
        if (repos != null) {
            return repos;
        }
        // cache update
        const response = await this.getPublicRepositories();
        repos = response.map(repo => ({
            name: repo.name,
            fullName: repo.full_name,
            url: repo.html_url,
            description: repo.description,
            language: repo.language,
            stars: repo.stargazers_count,
            forks: repo.forks_count,
            updatedAt: repo.updated_at
        }));
        cache.set(this.CACHE_KEY, repos);
        return repos;
    }

    private static async getPublicRepositories() {
        const url = `${this.BASE_URL}/users/${this.USERNAME}/repos?sort=updated&per_page=100`;
        const response = await axios.get<any[]>(url, {
            headers: {
                "Accept": "application/vnd.github+json",
                ...(this.GITHUB_TOKEN && { "Authorization": `Bearer ${this.GITHUB_TOKEN}` }),
                "User-Agent": "My-Portfolio"
            },
            timeout: 5000
        });
        return response.data;
    }

    private static getCache(): GithubRepo[] | null {
        return cache.get<GithubRepo[]>(this.CACHE_KEY) ?? null;
    }
}

File: backend/portfolio/src/main/java/com/example/portfolio/profile/service/ProfileService.java
package com.example.portfolio.profile.service;

import com.example.portfolio.profile.dto.ProfileDTO;
import com.example.portfolio.profile.dto.RepositoryDTO;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.core.io.support.ResourcePatternResolver;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class ProfileService {

    private ProfileDTO profile;
    private final Map<String, String> repoThumbnails = new HashMap<>();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final RedisService redisService;

    @Value("${github.token}")
    private String githubToken;
    @Value("${github.username}")
    private String username;
    @Value("${github.baseurl}")
    private String baseUrl;

    private final RestTemplate restTemplate;
    private final List<String> blackList = List.of("jung18", "Algorithm", "TIL");

    @PostConstruct
    public void loadProfile() {
        try {
            ClassPathResource profileJson = new ClassPathResource("data/portfolio.json");
            this.profile = objectMapper.readValue(
                    profileJson.getInputStream(),
                    new TypeReference<>() {}
            );
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    @PostConstruct
    public void loadThumbnails() {
        try {
            PathMatchingResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
            Resource[] resources = resolver.getResources("classpath:/static/thumbnails/*");
            for (Resource resource : resources) {
                if (resource.isReadable() && resource.getFilename() != null) {
                    repoThumbnails.put(resource.getFilename().split("\\.")[0], resource.getFilename());
                }
            }
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    public ProfileDTO getProfile() {
        if (profile != null) {
            return profile;
        }
        return new ProfileDTO();
    }

    public List<RepositoryDTO> getRepos() {
        List<RepositoryDTO> cache = redisService.getReposFromCache();
        if (cache != null) {
            return cache;
        }
        // new request -> cache update
        String apiUrl = baseUrl + "/users/" + username + "/repos?per_page=100";
        ResponseEntity<String> response = getGithubAPIResponse(apiUrl);

        List<RepositoryDTO> repositories = objectMapper.readValue(response.getBody(), new TypeReference<>() {});
        List<RepositoryDTO> dtos = repositories.stream()
                .filter(repo -> !blackList.contains(repo.getName()))
                .map(repo -> {
                    String thumbnail = repoThumbnails.get(repo.getName());
                    repo.setThumbnail(thumbnail);
                    return repo;
                }).toList();
        redisService.setReposToCache(dtos);
        return dtos;
    }

    private ResponseEntity<String> getGithubAPIResponse(String apiUrl) {
        HttpHeaders headers = new HttpHeaders();
        headers.set("Accept", "application/vnd.github+json");
        headers.set("Authorization", "Bearer " + githubToken);
        headers.set("User-Agent", "My-Portfolio");

        try {
            HttpEntity<String> httpEntity = new HttpEntity<>(headers);
            return restTemplate.exchange(
                    apiUrl,
                    HttpMethod.GET,
                    httpEntity,
                    String.class
            );
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    public List<String> getTechStacks() {
        try {
            ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
            Resource[] resources = resolver.getResources("classpath:/static/icons/*");
            List<String> fileNames = new ArrayList<>();

            for (Resource resource : resources) {
                if (resource.isReadable() && resource.getFilename() != null) {
                    String stackPath = "/icons/" + resource.getFilename();
                    fileNames.add(stackPath);
                }
            }
            return fileNames;
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }
}
